container-search/src/main/java/com/yahoo/search/dispatch/LoadBalancer.java

```diff
@@ -3,25 +3,108 @@
 import com.yahoo.search.Query;
 import com.yahoo.search.dispatch.SearchCluster.Group;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 public class LoadBalancer {
+    // The implementation here is a simplistic least queries in flight + round-robin load balancer
+    // TODO: consider the options in com.yahoo.vespa.model.content.TuningDispatch
 
-    private final SearchCluster searchCluster;
+    private final static Logger log = Logger.getLogger(LoadBalancer.class.getName());
+
+    private final boolean isInternallyDispatchable;
+    private final List<GroupSchedule> scoreboard;
+    private int needle = 0;
 
     public LoadBalancer(SearchCluster searchCluster) {
-        this.searchCluster = searchCluster;
+        if (searchCluster == null) {
+            this.isInternallyDispatchable = false;
+            this.scoreboard = null;
+            return;
+        }
+        this.isInternallyDispatchable = (searchCluster.groupSize() == 1);
+        this.scoreboard = new ArrayList<>(searchCluster.groups().size());
+
+        for (Group group : searchCluster.groups().values()) {
+            scoreboard.add(new GroupSchedule(group));
+        }
+        Collections.shuffle(scoreboard);
     }
 
     public Optional<Group> getGroupForQuery(Query query) {
-        if (searchCluster.groups().size() == 1) {
-            for(Group group: searchCluster.groups().values()) {
-                // since the number of groups is 1, this will run only once
-                if(group.nodes().size() == 1) {
-                    return Optional.of(group);
+        if (!isInternallyDispatchable) {
+            return Optional.empty();
+        }
+
+        return allocateNextGroup();
+    }
+
+    public void releaseGroup(Group group) {
+        synchronized (this) {
+            for (GroupSchedule sched : scoreboard) {
+                if (sched.group.id() == group.id()) {
+                    sched.adjustScore(-1);
+                    break;
+                }
+            }
+        }
+    }
+
+    private Optional<Group> allocateNextGroup() {
+        synchronized (this) {
+            GroupSchedule bestSchedule = null;
+
+            int index = needle;
```

Comment:
I found this a bit hard to read. Some suggestions:
- Create a ModuloInteger or something which encapsulates the wraparound - here you need to "manually" wrap around two different pointers.
- If you give that class a plus(int i) method which returns the wrapped-around result of adding i to it, you can get rid of index completely and just use needle.plus(i).
- I suggest a custom method returning boolean to compare schedules instead of compareTo, for example sched.preferredOver(other), then it can always be preferred over null as well.

Reply:
I'm replacing the wraparounds with a method at this time, and replacing the comparison logic with a boolean method. The result is much cleaner, as expected.